### PR TITLE
another attempt to fix netgame reload

### PR DESCRIPTION
**Pull request by rfomin**

### CODE REVIEW

File: Source/g_game.c [hunk 1]
```diff
@@ -714,12 +709,8 @@ static void G_ReloadLevel(void)
     G_BeginRecording();
   }
 
-  // force players to be initialized upon first level load
-  for (i = 0; i<MAXPLAYERS; i++)
-   if (playeringame[i])
-    players[i].playerstate = PST_REBORN;
-
-  G_DoLoadLevel();
+  G_InitNew(gameskill, gameepisode, gamemap);
```
Comment (fabiangreffrath) [inline Source/g_game.c:712]: Isn't this the route that resets `netgame` to `false`?

Comment (rfomin) [reply Source/g_game.c:712]: `netgame` is resets in `G_DoNewGame`. I tested this branch it seems to work.

**Comment by fabiangreffrath** [comment]
Calling out to @JCDorne and @Ch0wW.

**Comment by JCDorne** [comment]
We just tried this. We got a consistency failure when we reload the map (demo recording, complevel boom on plutonia map01, coop 2 players).

**Comment by rfomin** [comment]
> We just tried this. We got a consistency failure when we reload the map (demo recording, complevel boom on plutonia map01, coop 2 players).

Do you mean checksum failure? I cant reproduce it, it works without desync on my local network. At what point of this map you've got desync?

**Comment by fabiangreffrath** [comment]
Would be interesting to see what happens if we comment out the `I_Error()` line and just let the recordings go...

**Comment by fabiangreffrath** [comment]
Does this only happen on maps with Revenants? I mean, because after all, this could still be the Revenant basetic bug. What if this only happens if client and server are at least one gametic apart? This would also explain why it doesn't occur in local multiplayer games.

**Comment by rfomin** [comment]
I think in multiplayer game clients can't be one gametic apart or it would be a desync/consistency failure. And I think it doesn't matter if it's a LAN or the Internet, at least in theory.

**Comment by fabiangreffrath** [comment]
Seems I am out of ideas. 🤷

**Comment by JCDorne** [comment]
> At what point of this map you've got desync?

Right at the beginning, where the SuperShotgun is. Fighting the revenants. Close to 10 seconds.

> Would be interesting to see what happens if we comment out the `I_Error()` line and just let the recordings go...

I think that'll just lead to a desynced demo...

> Does this only happen on maps with Revenants? I mean, because after all, this could still be the Revenant basetic bug. What if this only happens if client and server are at least one gametic apart? This would also explain why it doesn't occur in local multiplayer games.

I don't know, We stopped testing right after that consistency failure...
The first demo in coop also desync in dsda-doom (near 1 minute mark).
I also tested -solo-net and the first demo also desync in dsda-doom

I think i'll recheck everything with a clean woof & dsda-doom folder. But I doubt that I did something wrong...

**Comment by rfomin** [comment]
> Right at the beginning, where the SuperShotgun is. Fighting the revenants. Close to 10 seconds.

That's really strange, I can reproduce it. My command line:
server: `woof -iwad ..\plutonia.wad -privateserver -complevel boom -record dem1`
client: `woof -autojoin`
new folder/config binary: https://github.com/fabiangreffrath/woof/suites/5903040919/artifacts/200999572

**Comment by JCDorne** [comment]
Just tried your build. Got a consistency failure at the same spot.
Server cmdline : `start woof -server -iwad plutonia.wad -complevel boom -warp 01 -nodes 2 -record DEMONAME -skill 4`
Client cmdline : `start woof.exe -iwad plutonia.wad -warp 01 -skill 4 -complevel boom -connect ip -record DEMONAME`
[plutoniacoopboom.zip](https://github.com/fabiangreffrath/woof/files/8404511/plutoniacoopboom.zip)
The first demo desync in dsdadoom (should finish the level in 56 seconds), and the second one stop because of the consistency failure.

**Comment by Ch0wW** [comment]
Additionally, we've observed that we never got any consistency failure if one of the players doesn't move on a reloaded demo. But the moment someone moves, it'll trigger a consistency failure at a random moment.

**Comment by rfomin** [comment]
You can switch between player views by pressing the spy key (F12 by default), so two demos are not needed. My only theory is that LAN and Internet are somehow different, it works for me on LAN.

**Comment by JCDorne** [comment]
Of course you can spy both POV in a demo.
But those two demos here are differents (but recorded at the same though). That why we provided both demos.
In the first one, the green player took damage. if you check the second demo file, the green player don't get any damage.
You can't see that with only one demofile. And that why it lead to a consistency failure.

Btw, we just tried doom2.wad. Map01 to Map03 (so no revenants at all), complevel boom, no reload.
The demo desync on map03.
So I think the revenant projectile problem isn't related.

> My only theory is that LAN and Internet are somehow different, it works for me on LAN.

Hmm, I tried with -solo-net and the first demo desync aswell.

**Comment by rfomin** [comment]
In [pluto_desync_reload.zip](https://github.com/fabiangreffrath/woof/files/8404522/pluto_desync_reload.zip)
player1_view.lmp  `rngseed = 1648989662`
player2_view.lmp  `rngseed = 1648989661`
@fabiangreffrath I assume we need to sync rngseed?

**Comment by fabiangreffrath** [comment]
Oh,wow, looks like this is indeed necessary. Will we have to send its value over network or will it be sufficient to uncouple it from `time()` and make it more deterministic?

**Comment by rfomin** [comment]
> will it be sufficient to uncouple it from `time()` and make it more deterministic?

This option is simpler. To be honest I'm not sure why `rngseed` is important, Choco/Crispy works fine without it.

**Comment by fabiangreffrath** [comment]
If this only affects reloaded multiplayer demo attempts, we could replace `rngseed = time(NULL)` with e.g `rngseed += gametics` in `G_ReloadLevel()`.

**Comment by rfomin** [comment]
@JCDorne and @Ch0wW, could you please test the latest commit? Binary: https://github.com/fabiangreffrath/woof/suites/5913007119/artifacts/201807536

**Comment by fabiangreffrath** [comment]
> Additionally, we've observed that we even got a consistency failure EVEN if one of the players doesn't move on a reloaded demo. (BOOM Complevel by the way).
> 
> Here's a demo showing what we had on our last desync: [pluto_desync_reload.zip](https://github.com/fabiangreffrath/woof/files/8404522/pluto_desync_reload.zip)
> 
> Player 1 never moved during that test, but Player 2 heard at one point Player 1 was hit, triggering the consistency failure.

To provide some rationale: The consistency error occurs if the movement of one of the players as reported by the clients doesn't match the movement as expected by the server based on the exchanged ticcmds. In this context "movement" is determined by the player's `x` coordinate, so not moving at all is the safest choice - until you're hit by an unexpected projectile, which will of course trigger the error.

**Comment by JCDorne** [comment]
> @JCDorne and @Ch0wW, could you please test the latest commit? Binary: https://github.com/fabiangreffrath/woof/suites/5913007119/artifacts/201807536

No more consistency failure :)
However, first demo desync and the reloaded demo aswell...

**Comment by fabiangreffrath** [comment]
> No more consistency failure :) 

Yes! 💪 

> However, first demo desync and the reloaded demo aswell...

Alright. Could you please add them here? Which complevel were you using?

**Comment by rfomin** [comment]
Now plutoniacoopboom-00000.lmp desync in the same way in Woof and PrBoom+.

**Comment by fabiangreffrath** [comment]
Awesome, thank you so much!

**Review by fabiangreffrath — APPROVED** [review]

**Comment by JCDorne** [comment]
Interesting, nice catch !
I'll try to record some demos today with those changes, in all complevel (and solonet) to check if everything is working :)

**Comment by JCDorne** [comment]
Okay, we just did some tests.
Look like boom complevel demos are fine so far.
But we tried mbf with Valiant and both demos desync (the first one and the reload)
[valiantcoopmbf.zip](https://github.com/fabiangreffrath/woof/files/8453230/valiantcoopmbf.zip)

**Comment by rfomin** [comment]
The desync in PrBoom+ in the first demo is due to this code:  https://github.com/coelckers/prboom-plus/blob/784a17d73818cffa4ec3114cef99bcc990d157f4/prboom2/src/p_inter.c#L758 I think this is a bug on PrBoom's part - `P_Random()` should not be called here. It also only happens in multiplayer. 
@fabiangreffrath what to do here? Copy all this logic in Woof?

**Comment by kraflab** [comment]
There is at least one mbf+ coop demo on dsda. So, as much as this code is a bit weird, probably it should be inherited by woof (or at least just make a P_Random call to keep things in sync 😄

**Comment by Ch0wW** [comment]
There is at least one mbf+ coop demo on dsda. So, as much as this code is a bit weird, probably it should be inherited by woof (or at least just make a P_Random call to keep things in sync :smile:

What demo are you referring to? [That 4P Nomo cooperative one](https://www.dsdarchive.com/files/demos/doom2/26949/lvallxco228.zip)? I've commented that line regarding that P_Random refered above on the latest DSDA-Doom release, and it playbacked fine. Do you have a specific list of demos of every multiplayer MBF+ complevel around?

**Comment by kraflab** [comment]
I don't have a list. Keep in mind that there will always be people that use old versions of ports, and I'd rather continue treating prboom+ as the standard for cl11, however odd this particular case may be.

**Comment by rfomin** [comment]
Patch:
```c
--- a/Source/p_inter.c
+++ b/Source/p_inter.c
@@ -692,6 +692,20 @@ static void P_KillMobj(mobj_t *source, mobj_t *target)
 	  if (!(target->flags & MF_FRIEND))
 	    players->killcount++;
         }
+#ifndef MBF_STRICT
+      // For compatibility with PrBoom+ complevel 11 and higher
+      else
+        if (demo_version >= 203 && !deathmatch &&
+             !(target->lastenemy &&
+               target->lastenemy->health > 0 &&
+               target->lastenemy->player
+              ) // Fighting a player
+            )
+          {
+            if (playeringame[1])
+              P_Random(pr_friends);
+          }
+#endif
 
   if (target->player)
     {
```
Or we can implement PrBoom+ monster kill counting logic.

**Comment by fabiangreffrath** [comment]
Oh my, PrBoom's strange netgame kill counting code is at it again. I once even got a pull request for Crispy implementing a part of it, but this has never come to fruition. 

Anyway, I agree with @kraflab here: If we want to achieve "complevel" compatibility we will have to adapt to PrBoom+ and all its quirks - even if this means to modify original MBF code for complevel 11 compatibility. 

That's why I like @rfomin's patch. It implements only the parts necessary to achieve compatibility with PrBoom+ and even guards the changes with the `MBF_STRICT` macro.

**Comment by fabiangreffrath** [comment]
So, this should be it, right?

**Comment by rfomin** [comment]
> So, this should be it, right?

Maybe. There are very few multiplayer demos available in the DSDA archive, and most of them are complevel vanilla, so I'm not sure. Hopefully we'll get more tests from the community.

**Comment by fabiangreffrath** [comment]
Okay, maybe we can get a final set if tests before I merge this. 😉

**Comment by JCDorne** [comment]
Oaky so, I just tested with @Ch0wW and so far, everything is working.
All demos sync with dsdadoom and even solonet is functional (reloaded or not, all complevel).

**Comment by fabiangreffrath** [comment]
Thanks to everyone involved!

**Pull request merged**